mlir/lib/CAPI/Dialect/Rock.cpp

```diff
@@ -0,0 +1,79 @@
+//===- MIGraphX.cpp - C Interface for MIGraphX dialect---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir-c/Dialect/Rock.h"
+#include "mlir/CAPI/IR.h"
+#include "mlir/CAPI/Pass.h"
+#include "mlir/CAPI/Registration.h"
+#include "mlir/CAPI/Wrap.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/Tuning/ConvContext.h"
+#include "mlir/Dialect/Rock/Tuning/RockTuning.h"
+
+#include <vector>
+
+MLIR_DEFINE_CAPI_DIALECT_REGISTRATION(Rock, rock, mlir::rock::RockDialect)
+DEFINE_C_API_PTR_METHODS(MlirRockTuningSpace, mlir::rock::TunableParams)
+DEFINE_C_API_PTR_METHODS(MlirRockTuningParam, mlir::rock::ParamEntry)
+// DEFINE_C_API_METHOD(MlirRockTuningKey, const void);
+
+using namespace mlir;
+
+MLIR_CAPI_EXPORTED MlirRockTuningSpace
+mlirRockTuningSpaceCreate(MlirModule module) {
+  struct rock::TunableParams *newParams;
+  auto mod = unwrap(module);
+  newParams = rock::createTunableParams(mod);
+  return wrap(newParams);
+}
+
+MLIR_CAPI_EXPORTED int
+mlirRockTuningGetNumParamsQuick(MlirRockTuningSpace params) {
+  auto tuningSpace = unwrap(params);
+  return tuningSpace->numHeuristicQuick;
+}
+
+MLIR_CAPI_EXPORTED
+int mlirRockTuningGetNumParamsFull(MlirRockTuningSpace params) {
+  auto tuningSpace = unwrap(params);
+  return tuningSpace->tuningRange.size();
+}
+
+// Allocate parameter entry to hold string allocation and pass it more easily.
+MLIR_CAPI_EXPORTED MlirRockTuningParam
+mlirRockTuningCreateParamAt(MlirRockTuningSpace params, int pos) {
+  auto tuningSpace = unwrap(params);
+  rock::ParamEntry *param = new rock::ParamEntry();
```

Comment:
Yeah, this is vaguely sus. So having the entire tuning space be an attribute might be the right call?

If nothing else, why not return a pointer into the vector?

Reply:
This is to have an own allocation of one params entry without tuning space. Later, I'd like to add another api creating ParamEntry from perfconfig string.

Reply:
https://mlir.llvm.org/docs/CAPI/#naming-convention-and-ownership-model
mlirXCreate : creates object and caller gets the ownership
and should be freed by - mlirXDestroy